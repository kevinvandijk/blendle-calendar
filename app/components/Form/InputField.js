import React from 'react';
import moment from 'moment';
import { PropTypes } from '../../helpers';

const { oneOf, string, func } = PropTypes;

class InputField extends React.Component {
  static propTypes = {
    label: string,
    name: string.isRequired,
    type: oneOf(['text', 'textarea', 'date']),
    value: string,
    onBlur: func,
    onFocus: func,
    onChange: func
  }

  static defaultProps = {
    type: 'text',
    value: ''
  }

  static contextTypes = {
    attach: func,
    detach: func
  }


  constructor(props) {
    super(props);

    this.state = {
      value: props.value
    };
  }

  state = {
    value: ''
  }

  componentWillMount() {
    if (this.context.attach) {
      this.context.attach(this);
    }
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.value !== this.props.value) {
      this.setState({
        value: nextProps.value
      });
    }
  }

  componentWillUnmount() {
    if (this.context.detach) {
      this.context.detach(this);
    }
  }

  onChange = (e) => {
    let value = e.target.value;
    if (this.props.type === 'date') {
      // When entering time allow only digits and : and .
      value = value.replace(/[^\d,:,.]+/gi, '');

      // When entering time notation with : and ., the max is 5 characters
      // When only entering digits, the max is 4 characters
      if (value.match(/\d{1,2}[:,.]/g)) {
        value = value.substr(0, 5);
      } else {
        value = value.substr(0, 4);
      }
    }

    this.setState({
      value
    });

    if (this.props.onChange) this.props.onChange(value);
  }

  onBlur = () => {
    this.setState({
      focused: false
    });

    if (this.props.type === 'date' && this.state.value.length) {
      const value = this.state.value.replace(/[^\d]/g, '');

      let formattedValue;
      if (value.length <= 2) {
        formattedValue = moment().hour(value).startOf('hour').format('HH:mm');
      } else {
        const minutes = value.slice(-2);
        const hours = value.slice(0, -2);

        formattedValue = moment().hour(hours).minute(minutes).format('HH:mm');
      }

      this.setState({
        value: formattedValue
      });
    }

    if (this.props.onBlur) this.props.onBlur(this.state.value);
  }

  onFocus = () => {
    this.setState({
      focused: true
    });

    if (this.props.onFocus) this.props.onFocus();
  }

  getValue() {
    return this.state.value;
  }

  setValue(value) {
    this.setState({
      value
    });
  }

  render() {
    let inputClass = 'Form-InputField';
    const { type, name, label } = this.props;
    if (type === 'date') inputClass = `${inputClass} Form-InputField-date`;
    if (type === 'textarea') inputClass = `${inputClass} Form-InputField-textarea`;

    return (
      <label htmlFor={ name } className="Form-InputField-label">
        { label }

        { type === 'textarea'
          ?
            <textarea
              id={ name }
              className={ inputClass }
              value={ this.state.value }
              onChange={ this.onChange }
              onBlur={ this.onBlur }
              onFocus={ this.onFocus }
              ref={ (ref) => { this.input = ref; } }
            />
          :
            <input
              id={ name }
              className={ inputClass }
              type="text"
              value={ this.state.value }
              onChange={ this.onChange }
              onBlur={ this.onBlur }
              onFocus={ this.onFocus }
              ref={ (ref) => { this.input = ref; } }
            />
        }
      </label>
    );
  }
}

export default InputField;
