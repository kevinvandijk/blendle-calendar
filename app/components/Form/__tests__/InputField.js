import React from 'react';
import renderer from 'react-test-renderer';
import stubContext from 'react-stub-context';
import { shallow, mount } from 'enzyme';
import InputField from '../InputField';

describe('Components', () => {
  describe('Form', () => {
    it('Renders an input field without a label', () => {
      const tree = renderer.create(
        <InputField name="name" />
      ).toJSON();

      expect(tree).toMatchSnapshot();
    });

    it('Renders an input field with a label when label is given', () => {
      const tree = renderer.create(
        <InputField name="name" label="The label yo" />
      ).toJSON();

      expect(tree).toMatchSnapshot();
    });

    it('Renders a date input when type is set to date', () => {
      const tree = renderer.create(
        <InputField name="name" type="date" />
      ).toJSON();

      expect(tree).toMatchSnapshot();
    });

    it('Renders a textarea when type is set to textarea', () => {
      const tree = renderer.create(
        <InputField name="name" type="textarea" />
      ).toJSON();

      expect(tree).toMatchSnapshot();
    });

    it('Returns the current value', () => {
      const value = 'Panda panda';
      const wrapper = shallow(<InputField name="name" value={ value } />);
      const instance = wrapper.instance();
      const initialResult = instance.getValue();

      expect(initialResult).toBe(value);

      const newValue = 'Panda?';
      wrapper.setState({ value: newValue });

      expect(instance.getValue()).toBe(newValue);
    });

    it('Sets a new value', () => {
      const value = 'Panda panda';
      const instance = shallow(<InputField name="name" />).instance();
      instance.setValue(value);

      const result = instance.getValue();
      expect(result).toBe(value);
    });

    it('Attaches to a form on mount when it has an attach method available within context', () => {
      const attachStub = jest.fn();
      const InputFieldWithContext = stubContext(InputField, {
        attach: attachStub
      });

      mount(<InputFieldWithContext name="bla" />);

      expect(attachStub).toHaveBeenCalled();
    });

    it('Detaches from a form on unmount when it has a detach method available within context', () => {
      const detachStub = jest.fn();
      const InputFieldWithContext = stubContext(InputField, {
        detach: detachStub
      });

      const wrapper = mount(<InputFieldWithContext name="bla" />);
      wrapper.unmount();

      expect(detachStub).toHaveBeenCalled();
    });
  });
});
