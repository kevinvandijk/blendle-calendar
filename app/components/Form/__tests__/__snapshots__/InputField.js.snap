exports[`Components Form Renders a date input when type is set to date 1`] = `
<label
  className="Form-InputField-label"
  htmlFor="name">
  <input
    className="Form-InputField Form-InputField-date"
    id="name"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    onKeyUp={[Function]}
    type="text"
    value="" />
</label>
`;

exports[`Components Form Renders a textarea when type is set to textarea 1`] = `
<label
  className="Form-InputField-label"
  htmlFor="name">
  <textarea
    className="Form-InputField Form-InputField-textarea"
    id="name"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    onKeyUp={[Function]}
    value="" />
</label>
`;

exports[`Components Form Renders an input field with a label when label is given 1`] = `
<label
  className="Form-InputField-label"
  htmlFor="name">
  The label yo
  <input
    className="Form-InputField"
    id="name"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    onKeyUp={[Function]}
    type="text"
    value="" />
</label>
`;

exports[`Components Form Renders an input field without a label 1`] = `
<label
  className="Form-InputField-label"
  htmlFor="name">
  <input
    className="Form-InputField"
    id="name"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    onKeyUp={[Function]}
    type="text"
    value="" />
</label>
`;
